toolchain/semantics/semantics_ir.cpp

```diff
@@ -399,4 +403,80 @@ auto SemanticsIR::StringifyType(SemanticsTypeId type_id) -> std::string {
   return str;
 }
 
+auto GetSemanticsExpressionCategory(const SemanticsIR& semantics_ir,
+                                    SemanticsNodeId node_id)
+    -> SemanticsExpressionCategory {
+  const SemanticsIR* ir = &semantics_ir;
+  while (true) {
+    auto node = ir->GetNode(node_id);
+    switch (node.kind()) {
+      case SemanticsNodeKind::Invalid:
+      case SemanticsNodeKind::Assign:
+      case SemanticsNodeKind::Branch:
+      case SemanticsNodeKind::BranchIf:
+      case SemanticsNodeKind::BranchWithArg:
+      case SemanticsNodeKind::FunctionDeclaration:
+      case SemanticsNodeKind::Namespace:
+      case SemanticsNodeKind::Return:
+      case SemanticsNodeKind::ReturnExpression:
+      case SemanticsNodeKind::StructTypeField:
+        return SemanticsExpressionCategory::NotExpression;
+
+      case SemanticsNodeKind::CrossReference: {
+        auto [xref_id, xref_node_id] = node.GetAsCrossReference();
+        ir = &semantics_ir.GetCrossReferenceIR(xref_id);
+        node_id = xref_node_id;
+        continue;
```

Comment:
Fun implementation of recursion. ;]

This is just to help me learn -- what about cross references prevents this from being an infinite loop?

Reply:
We only build cross references to the IRs for our dependencies, and our dependency graph is acyclic.